subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/ValidationActions.java

```diff
@@ -110,9 +113,81 @@ public void doValidate(String propertyName, Object values, TaskValidationContext
+    private static void reportMissingInput(TaskValidationContext context, String kind, String propertyName, File input) {
+        context.visitPropertyProblem(problem -> {
+            String lowerKind = kind.toLowerCase();
+            problem.withId(ValidationProblemId.INPUT_DOES_NOT_EXIST)
+                .forProperty(propertyName)
+                .reportAs(ERROR)
+                .withDescription(() -> lowerKind + " '" + input + "' doesn't exist")
+                .happensBecause("An input is missing")
+                .addPossibleSolution(() -> "Make sure the " + lowerKind + " exists before the task is called")
+                .addPossibleSolution(() -> "Make sure that the task which produces the " + lowerKind + " is declared as an input")
+                .documentedAt("validation_problems", "input_does_not_exist");
+        });
+    }
+
+    private static void reportUnexpectedInputKind(TaskValidationContext context, String kind, String propertyName, File input) {
+        context.visitPropertyProblem(problem -> {
+            String lowerKind = kind.toLowerCase();
+            problem.withId(ValidationProblemId.UNEXPECTED_INPUT_TYPE)
+                .forProperty(propertyName)
+                .reportAs(ERROR)
+                .withDescription(() -> lowerKind + " '" + input + "' is not a " + lowerKind)
+                .happensBecause(() -> "Expected an input to be a " + kind.toLowerCase() + " but it was a " + actualKindOf(input))
+                .addPossibleSolution(() -> "Use a " + lowerKind + " as an input")
+                .addPossibleSolution(() -> "Declare the input as a " + actualKindOf(input) + " instead")
+                .documentedAt("validation_problems", "unexpected_input_type");
+        });
+    }
+
+    private static void reportCannotWriteToDirectory(String propertyName, TaskValidationContext context, File directory, String cause) {
+        context.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
+                .reportAs(ERROR)
+                .forProperty(propertyName)
+                .withDescription(() -> "is not writable because " + cause)
+                .happensBecause(() -> "Expected '" + directory + "' to be a directory but it's a " + actualKindOf(directory))
+                .addPossibleSolution("Make sure that the '" + propertyName + "' is configured to a directory")
+                .documentedAt("validation_problems", "cannot_write_output")
+        );
+    }
+
+
+    private static void reportCannotWriteToFile(String propertyName, TaskValidationContext context, String cause) {
+        context.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
+                .reportAs(ERROR)
+                .forProperty(propertyName)
+                .withDescription(() -> "is not writable because " + cause)
+                .happensBecause(() -> "Cannot write a file to a location pointing at a directory")
```

Comment:
This is strange. The `happensBecause` seems to be in the description, right? Maybe something like:

```suggestion
                .withDescription(() -> "cannot be used as a file output")
                .happensBecause(() -> "It is not writable because " + cause)
```

Reply:
Not really: the description is the first thing displayed, so it gives information about the problem, like "cannot write Foo" (the _what_). Then the happens because says "and this happened because..." and tries to explain _why_ this problem happens (not _what_). Sometimes it's a bit redundant information, I agree.